File: solutions/php/bob/1/Bob.php
<?php

declare(strict_types=1);

class Bob
{
    public function respondTo(string $str): string
    {
        // Remove whitespace and check for silence
        $trimmed = trim($str);
        
        if (empty($trimmed)) {
            return "Fine. Be that way!";
        }
        
        // Check if it's yelling (all uppercase and has letters)
        $isYelling = $this->isYelling($trimmed);
        
        // Check if it's a question (ends with question mark)
        $isQuestion = substr($trimmed, -1) === '?';
        
        if ($isYelling && $isQuestion) {
            return "Calm down, I know what I'm doing!";
        }
        
        if ($isYelling) {
            return "Whoa, chill out!";
        }
        
        if ($isQuestion) {
            return "Sure.";
        }
        
        return "Whatever.";
    }
    
    private function isYelling(string $str): bool
    {
        // Check if string contains letters and all letters are uppercase
        $hasLetters = preg_match('/[a-zA-Z]/', $str);
        $isAllUppercase = $str === strtoupper($str);
        
        return $hasLetters && $isAllUppercase;
    }
}

File: solutions/ruby/bob/1/bob.rb
class Bob
  def self.hey(remark)
    # Limpiamos espacios en blanco innecesarios
    remark = remark.strip

    # 1. ¿No dijo nada?
    return "Fine. Be that way!" if remark.empty?

    # Definimos estados para mayor claridad
    is_question = remark.end_with?('?')
    # Un grito debe estar en mayúsculas y tener al menos una letra
    is_shouting = (remark == remark.upcase) && remark.match?(/[A-Z]/)

    # 2. Lógica de respuesta de Bob
    if is_shouting && is_question
      "Calm down, I know what I'm doing!"
    elsif is_shouting
      "Whoa, chill out!"
    elsif is_question
      "Sure."
    else
      "Whatever."
    end
  end
end